File: packages/domain/src/value-objects/quantity.ts
import { InvalidQuantityError } from '../errors/domain-error.js';

/** Positive integer quantity for sale lines and stock. */
export class Quantity {
  private constructor(readonly value: number) {}

  static of(value: number): Quantity {
    if (!Number.isInteger(value) || value < 1) {
      throw new InvalidQuantityError();
    }
    return new Quantity(value);
  }
}

File: backend/crates/domain-inventory/src/quantity.rs
use crate::error::InventoryError;

/// Positive integer quantity for sale lines and stock.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Quantity(i32);

impl Quantity {
    pub fn of(value: i32) -> Result<Self, InventoryError> {
        if value < 1 {
            return Err(InventoryError::InvalidQuantity);
        }
        Ok(Self(value))
    }

    pub fn value(self) -> i32 {
        self.0
    }
}
